### PR TITLE
feat: Add ComfyUI tool for Stable Diffusion

### PR DESCRIPTION
**Pull request by QunBB**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Fixes 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

img-1
img-2
img-3

### CODE REVIEW

File: api/core/tools/provider/builtin/comfyui/_assets/icon.png
Comment (crazywoola) [inline api/core/tools/provider/builtin/comfyui/_assets/icon.png:1]: https://storage.googleapis.com/comfy-assets/logo.png

The logo should be this one.

File: api/core/tools/provider/builtin/comfyui/comfyui.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  description:
+    en_US: ComfyUI is a tool for generating images which can be deployed locally.
+    zh_Hans: ComfyUI 是一个可以在本地部署的图片生成的工具。
+    pt_BR: ComfyUI is a tool for generating images which can be deployed locally.
+  icon: icon.png
+  tags:
+    - image
+credentials_for_provider:
+  base_url:
+    type: secret-input
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/comfyui/comfyui.yaml]: This does not need to be a `secret-input`.

File: api/core/tools/provider/builtin/comfyui/tools/txt2img.json [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      ]
+    },
+    "class_type": "KSampler",
+    "_meta": {
+      "title": "KSampler"
+    }
+  },
+  "4": {
+    "inputs": {
+      "ckpt_name": "3D角色PLUS+IP设计_v1.0.safetensors"
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/comfyui/tools/txt2img.json]: Please use English here.

**Review by crazywoola — CHANGES_REQUESTED** [review]
See comments

File: api/core/tools/provider/builtin/comfyui/comfyui.yaml (hunk 1 above)
Comment (QunBB) [reply api/core/tools/provider/builtin/comfyui/comfyui.yaml]: I has used `text-input` instead of `secret-input`.

**Comment by QunBB** [comment]
@crazywoola Done. Check it again please.

File: api/core/tools/provider/builtin/comfyui/comfyui.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+identity:
+  author: Dify
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/comfyui/comfyui.yaml]: You can leave your name here.

**Review by crazywoola — COMMENTED** [review]
See the author

File: api/core/tools/provider/builtin/comfyui/comfyui.yaml (hunk 3 above)
Comment (QunBB) [reply api/core/tools/provider/builtin/comfyui/comfyui.yaml]: Oh! Thanks

**Review by crazywoola — APPROVED** [review]

File: api/core/tools/provider/builtin/comfyui/tools/comfyui_stable_diffusion.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    def validate_models(self) -> Union[ToolInvokeMessage, list[ToolInvokeMessage]]:
+        """
+            validate models
+        """
+        try:
+            base_url = self.runtime.credentials.get('base_url', None)
+            if not base_url:
+                raise ToolProviderCredentialValidationError('Please input base_url')
+            model = self.runtime.credentials.get('model', None)
+            if not model:
```
Comment (crazywoola) [inline api/core/tools/provider/builtin/comfyui/tools/comfyui_stable_diffusion.py:175]: image
It seems this input is missing?

**Review by crazywoola — CHANGES_REQUESTED** [review]
See comments

**Comment by WeepsDanky** [comment]
Hi @QunBB . When I was testing this tools, an error shows up: Failed to get models, Please input model. My comfyui is started on http://127.0.0.1:8188/ with a blank canvas. How to resolve this error? 

image

**Comment by QunBB** [comment]
@crazywoola @WeepsDanky The model name should be also exposed when verifying tool's credentials. Some modified code was after verifying credentials, so i miss the test about the changed code about `validate_models`. I will fix it later.

**Comment by QunBB** [comment]
@crazywoola @WeepsDanky I have fixed it, please check it again.

**Comment by QunBB** [comment]
@crazywoola But i am troubled by Ruff check, should i reformat the files?

**Comment by WeepsDanky** [comment]
@QunBB Great thanks, it is working now.

**Comment by WeepsDanky** [comment]
@QunBB I noticed the tool currently can only use a pre-defined workflow txt2img.json. Can you please specify this in the description in yaml? We need to make sure other users understand this tool can only use this workflow. 

8f0d79af27fadd2f30849a0e849a65c

**Comment by QunBB** [comment]
@WeepsDanky Hi, i have change the name and description in yaml. In addition, i add the support for SD3 and FLUX.
You could try them like the examples: 
* [SD3](https://comfyanonymous.github.io/ComfyUI_examples/sd3/)
* [flux-fp8](https://comfyanonymous.github.io/ComfyUI_examples/flux/#simple-to-use-fp8-checkpoint-version)

File: api/core/tools/provider/builtin/comfyui/tools/comfyui_stable_diffusion.py (hunk 4 above)
Comment (QunBB) [reply api/core/tools/provider/builtin/comfyui/tools/comfyui_stable_diffusion.py:175]: You should input `http://127.0.0.1:8188`

**Review by laipz8200 — APPROVED** [review]

**Review by crazywoola — APPROVED** [review]

**Comment by wisepmlin** [comment]
Mac comfloxy
 Failed to get models, [Errno 111] Connection refused
截屏2024-09-21 14 06 22

**Comment by QunBB** [comment]
@wisepmlin This seems to be a network issue. I have tried to install Mac comflowy and it worked too.

**Comment by Dongnc1017** [comment]
https://docs.dify.ai/tutorials/tool-configuration/comfyui 的链接无法访问
1、请问我想使用flux 的模型，Flux Dev和Flux Schnell，都支持吗？是否下载过来放在ComfyUl/models/unet/ 就行了。
2、但是又说支持包含文本编码器/clip的SD1.5、SDXL、SD3和FLUX，但不支持需要clip加载器的模型。
这句话是什么意思？ 比如t5xxl fp16.safetensors 或t5xxl fp8 e4m3fn.safetensors:？
能否出一个详细的教程

**Comment by QunBB** [comment]
@Dongnc1017 The comments above have already mentioned it. You should download those models which contain text encoders, then put them into `ComfyUl/models/checkpoints/` like official comfyui examples:
* [SD3](https://comfyanonymous.github.io/ComfyUI_examples/sd3/)
* [flux-fp8](https://comfyanonymous.github.io/ComfyUI_examples/flux/#simple-to-use-fp8-checkpoint-version)

Flux Dev and Flux Schnell are both supported.

**Comment by laipz8200** [comment]
Hi @QunBB, would you be interested in updating the documentation for this tool? Or perhaps removing the inaccessible links from the configuration?

**Comment by QunBB** [comment]
@laipz8200 I'm intersted in updating it when i'm free later. I would replace the current link with the official ComfyUI website first, and will add it again when i finish the documentation in dify.

**Comment by hjlarry** [comment]
Hi  @QunBB   I think current ComfyUI tool can be better by reduce the user input and selection, just give the user a `prompt_text` input.

The comfyUi workflow is similar to dify's workflow, support  a variety of nodes.  Current implementation seems define specific steps, the user can only select each params of these steps, seems inflexible.

A more reasonable user case is: the user edit the workflow in the comfy UI, and then export the json of this workflow which is simliar to dify's DSL,  then paste it to this tool to get the image in dify.  Now they combine the dify and comfyUI's workflow.

This is the export API button:
image


This is a simple workflow just to change an ICON's style, but it has 40 nodes,  current solution can't do this:
image


what do you guys think about this ?

**Comment by QunBB** [comment]
@hjlarry Sure, we could add it, then it will support any image generation workflows in ComfyUI. But i don't think LLM could correctly generate the prompt text via agent, so it may only be used in dify's workflow.
By the way, the current tool could be continue to be used for beginners or for simplicity like the pattern of stable diffusion tool. And it could be both used in dify's agent and workflow.

**Comment by hjlarry** [comment]
@QunBB Sure, we can add a new tool and keep both of them.
I think the new tool can config the workflow's json and which node is the text node,  to make LLM just gen text prompts then used to agent app.

**Comment by QunBB** [comment]
@hjlarry It is a good idea.

**Comment by Kimdow220807** [comment]
Hi,This is the content of files output after generation. It seems that the image link is incomplete? How should I get the generated image?
"files": [
    {
      "__variant": "FileVar",
      "tenant_id": "241db99a-1ab2-408b-b2de-19a3ad342be1",
      "type": "image",
      "transfer_method": "tool_file",
      "url": "/files/tools/18a82aa7-ff46-4fd0-b6a9-868862938f3a.png?timestamp=1728357260&nonce=762e3c75ad0273a5fdd9569e0c5fe01a&sign=Jpn3yvvHe-81JwsjS_jWNTwRpwzq7dw1jSJwDeOOVgQ=",
      "remote_url": "/files/tools/18a82aa7-ff46-4fd0-b6a9-868862938f3a.png",
      "related_id": "18a82aa7-ff46-4fd0-b6a9-868862938f3a",
      "filename": "image",
      "extension": ".png",
      "mime_type": "image/png"
    }

**Pull request merged**